Html api/tag processor self closing flag

dmsnell opened the pull request:
Introduce has_self_closing_flag() to the HTML Tag Processor, exposing whether a currently-matched tag contains the self-closing flag /, used when determining if one needs to look for a closing tag of the same name.

This enables structural parsing of an HTML document and is needed to identify self-closing HTML foreign elements.

Trac ticket [#58009](https://core.trac.wordpress.org/ticket/58009#ticket)

cc: @adamziel @gziolo @ockham

ockham commented on src/wp-includes/html-api/class-wp-html-tag-processor.php:
```diff
@@ -1743,6 +1743,31 @@ public function get_tag() {
 	}
 
+	/**
+	 * Indicates if the currently matched tag contains the self-closing flag.
```
Maybe 
```suggestion
	 * Indicates if the currently matched tag contains the <self-closing/> flag.
```

to make it a bit more illustrative? 🤔

dmsnell replied:
was trying to follow the explicit constraint not to use HTML tags in summaries and tags

> No HTML markup or Markdown of any kind should be used in the summary. If the text refers to an HTML element or tag, then it should be written as “image tag” or “img” element, not “<img>“. - [PHPDoc standards](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#summary)

ockham commented on tests/phpunit/tests/html-api/wpHtmlTagProcessor.php:
```diff
@@ -51,6 +51,54 @@ public function test_get_tag_returns_open_tag_name() {
+	 *
+	 * @dataProvider data_has_self_closing_flag
+	 *
+	 * @param string $html input HTML whose first tag might contain the self-closing flag `/`.
+	 * @param bool $flag_is_set whether the input HTML's first tag contains the self-closing flag.
```
Maybe start the param descriptions with a capital letter? Might help legibility to delineate the description from the arg name. Also, AFAICS, all the other `@param` descriptions in this file do.

```suggestion
	 * @param string $html Input HTML whose first tag might contain the self-closing flag `/`.
	 * @param bool $flag_is_set Whether the input HTML's first tag contains the self-closing flag.
```

ockham reviewed (approved):
Thank you @dmsnell! I left two minor (non-blocking) notes. LGTM otherwise! 😄

adamziel commented on tests/phpunit/tests/html-api/wpHtmlTagProcessor.php:
```diff
@@ -51,6 +51,54 @@ public function test_get_tag_returns_open_tag_name() {
+	/**
+	 * Data provider. HTML tags which might have a self-closing flag, and an indicator if they do.
+	 *
+	 * @return array[]
```
Nit: 
```suggestion
	 * @return array[].
```

dmsnell replied:
is this even correct? most in the file don't have that period, and I would expect a period on a description, not following a type.

in either case, it's clear we need to clean up the consistency of it in the file. mind if I file a follow-up PR to do that all at one time?

adamziel commented on tests/phpunit/tests/html-api/wpHtmlTagProcessor.php:
```diff
@@ -51,6 +51,54 @@ public function test_get_tag_returns_open_tag_name() {
+		return array(
+			// These should not have a self-closer, and will leave an element un-closed if it's assumed they are self-closing.
+			'Self-closing flag on non-void HTML element' => array( '<div />', true ),
+			'No self-closing flag on non-void HTML element' => array( '<div>', false ),
```
It's interesting the linter doesn't complain about a misaligned arrow 🤔  I hate to be that person, but all the arrows need to be at the same offset.

dmsnell replied:
then don't be that person 😄 

I'll try this out, but I think the linter complains here if we line up the arrows. it has some line-length threshold at which it separates the alignment of shorter lines and longer lines.

dmsnell replied:
confirmed: the alignment that looks right is rejected because it doesn't follow the specific linting rules. now go away and take your legibility with you! 🙃

adamziel commented on tests/phpunit/tests/html-api/wpHtmlTagProcessor.php:
```diff
@@ -51,6 +51,54 @@ public function test_get_tag_returns_open_tag_name() {
+			'Self-closing flag after extra spaces'       => array( '<div      />', true ),
+			'Self-closing flag after attribute'          => array( '<div id=test/>', true ),
+			'Self-closing flag after quoted attribute'   => array( '<div id="test"/>', true ),
+			'Self-closing flag after boolean attribute'  => array( '<div enabled/>', true ),
```
Would you mind adding two more tests? I'm thinking of:

* `<img/>` <- without a space between the tag name and a closer
* `<img / >` <- `/` is actually an HTML attribute because of the trailing space.

dmsnell replied:
gladly! these are great test cases to include, and thankfully it handles it right already.

adamziel commented:
LGTM! I just left a few minor notes.

dmsnell committed:
HTML API: Add `has_self_closing_flag()` to Tag Processor

Trac ticket: [#58009](https://core.trac.wordpress.org/ticket/58009#ticket)

In this patch we're adding `has_self_closing_flag()` to the HTML Tag Processor.
This exposes whether a currently-matched tag contains the self-closing flag `/`.

This information is critical for the evolution of the HTML API in order
to track and parse HTML structure, specifically, knowing whether an
HTML foreign element is self-closing or not.

ockham commented:
Committed to Core `trunk` in [r55619](https://core.trac.wordpress.org/changeset/55619).
